### PR TITLE
Introduce sample application (spring-petclinic + postgresql) to be used in quickstart

### PR DESCRIPTION
**Pull request by pmacik**
### Motivation

Current quickstart uses only REST API server part of the Spring Petclinic as an application and Crunchy Postgresql Operator backed DB. It has couple of UX issues.

The current application:
* does not have UI so user must use `kubectl port-forward` and `curl` command to send requests to the application's API server and analyze responses for accessing and verifying that the application works as expected
* user has to manually run a script to initialize database with application schema and data (as one of the quickstart's steps)

The current database:
* requires installation of Crunchy Postgresql Operator (via OLM) that wouldn't be possible for user with kubernetes cluster without OLM
* uses an operator that is not available on other then `x86_64` architectures, which excludes all users on alternative architectures

### Changes

This PR:
* Introduces an alternative Sprint Petclinic application that improves UX:
    * has a UI for user to interact with
    * initializes database schema and data automatically, so user does not to do it manually - thus simlify the quickstart scenario
    * is based on https://github.com/spring-projects/spring-petclinic application enhanced by https://github.com/spring-cloud/spring-cloud-bindings library.
    * is multi-arch
* Introduces an alternative PostgreSQL deployment that:
    * is in a for of plain `Deployment` based on `postgres:13` image that is multi-arch
    * does not require OLM
* Both app and db including binding is tested on `minikube` and `openshift` clusters
* Is followed-up by https://github.com/redhat-developer/service-binding-operator/pull/1085
    * These app and db deployments are supposed to replace the ones used in SBO's quickstart eventually.

### Testing

The following steps involve running some convenient `Makefile` targets that serve here only to ease testing and will be replaced by the actual `kubectl` or `oc` commands and explained in the quickstart, eventually.

0. Set your environment
    * with minikube (default)
    `export KUBERNETES_RUNTIME=minikube`
    * with OpenShift
    `export KUBERNETES_RUNTIME=openshift`

1. Go to `samples/app/spring-petclinic` directory
`cd samples/app/spring-petclinic`

2. Create user namespace (`my-postgresql` by default)
`make namespace`

3. Deploy database Deployment
`make deploy-database`

4. Expose DB binding data by setting annotations on database Deployment
`make annotate-database`

5. Deploy application Deployment
`make deploy-app`
URL for the application UI is printed out so open it to access the application, but application's pod is stuck in `Init` state due unavailable DB.

6. Bind application to db
`make bind`

7. After the application pod is restarted, the application's pod is `Running` UI should show the web interface.

### CODE REVIEW

**Comment by pmacik** [comment]
@yselkowitz could you please take a look in terms of multi-arch? thx

File: samples/apps/spring-petclinic/.patch/application-postgresql.properties.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+index 80bd3e1..0000000
+--- a/src/main/resources/application-postgres.properties
++++ /dev/null
+@@ -1,6 +0,0 @@
+-database=postgres
+-spring.datasource.url=${POSTGRES_URL:jdbc:postgresql://localhost/petclinic}
+-spring.datasource.username=${POSTGRES_USER:petclinic}
+-spring.datasource.password=${POSTGRES_PASS:petclinic}
+-# SQL is written to be idempotent so this is safe
+-spring.datasource.initialization-mode=always
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.patch/application-postgresql.properties.patch:12]: why did we drop these lines? If I understand correctly here: https://docs.spring.io/spring-boot/docs/2.1.0.M1/reference/html/howto-database-initialization.html#howto-database-initialization 

> Spring Boot automatically creates the schema of an embedded DataSource. This behaviour can be customized by using the spring.datasource.initialization-mode property. For instance, if you want to always initialize the DataSource regardless of its type:
>
> spring.datasource.initialization-mode=always

Is this something what we want actually to happen and with that we can drop that initiContainer?

File: samples/apps/spring-petclinic/.patch/pom.xml.patch [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ 
+   <dependencies>
+@@ -46,6 +48,10 @@
+       <groupId>org.springframework.boot</groupId>
+       <artifactId>spring-boot-starter-data-jpa</artifactId>
+     </dependency>
++    <dependency>
++      <groupId>org.springframework.boot</groupId>
++      <artifactId>spring-boot-starter-jdbc</artifactId>
++    </dependency>
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.patch/pom.xml.patch:21]: I thought that we need only to add `spring-cloud-bindings` lib? Where do we need this?

File: samples/apps/spring-petclinic/.patch/pom.xml.patch [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     </dependency>
++    <dependency>
++      <groupId>org.springframework.data</groupId>
++      <artifactId>spring-data-jdbc-core</artifactId>
++      <version>${spring-data-jdbc.version}</version>
++      <exclusions>
++        <exclusion>
++          <groupId>org.springframework</groupId>
++          <artifactId>*</artifactId>
++        </exclusion>
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.patch/pom.xml.patch:49]: same question as above?

File: samples/apps/spring-petclinic/.patch/pom.xml.patch [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+       </snapshots>
+     </repository>
++    <repository>
++      <snapshots>
++        <enabled>false</enabled>
++      </snapshots>
++      <id>spring-libs-private-release</id>
++      <name>libs-private-release</name>
++      <url>https://repo.spring.io/artifactory/libs-private-release</url>
++    </repository>
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.patch/pom.xml.patch:66]: why do we need this?

File: samples/apps/spring-petclinic/Dockerfile.app [hunk 5]
```diff
@@ -0,0 +1,25 @@
+FROM maven:3.8.4-openjdk-11 as builder
+
+ARG PETCLINIC_REPO=https://github.com/spring-projects/spring-petclinic.git
+ARG PETCLINIC_COMMIT=ce626da705302e71771b7e22429d1cc3209e7ebb
+
+ENV PETCLINIC_DIR /workspace/petclinic.git
+
+WORKDIR ${PETCLINIC_DIR}
+
+RUN git clone ${PETCLINIC_REPO} ${PETCLINIC_DIR} && \
```
Comment (pedjak) [inline samples/apps/spring-petclinic/Dockerfile.app:10]: It seems to me that the code is checked out in `${PETCLINIC_DIR}/${PETCLINIC_DIR}` ?

File: samples/apps/spring-petclinic/db-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: spring-petclinic-postgresql
+stringData:
+  database: petclinic
+  username: petclinic
```
Comment (pedjak) [inline samples/apps/spring-petclinic/db-deployment.yaml:8]: since these values are same as those set in `application-postgres.properties` file, how do we know that they are really consumed from binding projection?

File: samples/apps/spring-petclinic/init-database.sh [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        sleep infinity
+    else
+        export PG_HOST=${PG_HOST:-$(cat $BINDING/host)}
+        export PG_PORT=${PG_PORT:-$(cat $BINDING/port)}
+        export PG_NAME=${PG_NAME:-$(cat $BINDING/database)}
+        export PG_USER=${PG_USER:-$(cat $BINDING/username)}
+        export PGPASSWORD=${PGPASSWORD:-$(cat $BINDING/password)}
+
+        if [[ -z "$(psql -h $PG_HOST -p $PG_PORT -U "$PG_USER" "$PG_NAME" -c "SELECT tablename FROM pg_catalog.pg_tables" | grep pets)" ]]; then
+            curl -LO https://raw.githubusercontent.com/spring-petclinic/spring-petclinic-rest/master/src/main/resources/db/postgresql/initDB.sql
```
Comment (pedjak) [inline samples/apps/spring-petclinic/init-database.sh:20]: if we really need to keep this init container (see my comment about auto schema creation above), we should download `.sql` files during container build process and have them in container ready, or even better commit them into our repo.. because the version on `spring-petclinic-rest` master branch can change over time and break us at the runtime.

File: samples/apps/spring-petclinic/.minikube/deploy-app.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+
+NAMESPACE=${NAMESPACE:-my-postgresql}
+
+sed -e 's,REPLACE_PETCLINIC_APP_IMAGE,'${PETCLINIC_APP_IMAGE}',g' app-deployment.yaml | \
+sed -e 's,REPLACE_PETCLINIC_APP_DBINIT_IMAGE,'${PETCLINIC_APP_DBINIT_IMAGE}',g' | \
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.minikube/deploy-app.sh:6]: the rest of the project uses `kustomize`, so if really needed we could use that instead of `sed`

File: samples/apps/spring-petclinic/.minikube/deploy-app.sh [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+kubectl apply -f - -n ${NAMESPACE} --wait
+kubectl delete service spring-petclinic --ignore-not-found=true -n ${NAMESPACE}
```
Comment (pedjak) [inline samples/apps/spring-petclinic/.minikube/deploy-app.sh:8]: why is this needed?

File: samples/apps/spring-petclinic/Makefile [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+namespace:
+ifeq ($(KUBERNETES_RUNTIME), openshift)
+	-oc new-project $(NAMESPACE)
+else ifeq ($(KUBERNETES_RUNTIME), minikube)
+	-kubectl create namespace $(NAMESPACE)
+	kubectl config set-context --current --namespace=$(NAMESPACE)
+endif
+
+.PHONY: install-sbo
+install-sbo:
```
Comment (pedjak) [inline samples/apps/spring-petclinic/Makefile:21]: do we really need to have here a target where we install SBO? If needed, we can refer directly `install.sh` from the repo itself, and not download it from `master` branch.

**Review by pedjak — CHANGES_REQUESTED** [review]
additionally, please avoid folder names starting with a dot, e.g. `.minikube`, `.openshift`.. they are hidden by default from shell and this can create some confusions.

File: samples/apps/spring-petclinic/Dockerfile.app (hunk 5 above)
Comment (sadlerap) [reply samples/apps/spring-petclinic/Dockerfile.app:10]: Since `${PETCLINIC_DIR}` is an absolute path (it starts with `/`), `git clone` will clone into `${PETCLINIC_DIR}` (`/workspace/petclinic.git`), not `${PETCLINIC_DIR}/${PETCLINIC_DIR}`.

File: samples/apps/spring-petclinic/.patch/pom.xml.patch (hunk 4 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/.patch/pom.xml.patch:66]: This maven repo is where `spring-cloud-binding` library is available

Comment (pedjak) [reply samples/apps/spring-petclinic/.patch/pom.xml.patch:66]: hm, it sounds like some private repo, not official one.. how about that we use `https://repo.spring.io/releases` instead?

File: samples/apps/spring-petclinic/.patch/pom.xml.patch (hunk 3 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/.patch/pom.xml.patch]: this was needed for `spring-data-jpa` profile used - not needed anymore

File: samples/apps/spring-petclinic/.patch/pom.xml.patch (hunk 4 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/.patch/pom.xml.patch]: Good catch! I missed that, I assumed the release repo was already there so i looked for it elsewhere. But it wasn't. I'll use the release repo instead. thx

File: samples/apps/spring-petclinic/db-deployment.yaml (hunk 6 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/db-deployment.yaml]: good point, I've change changed it to someting different to be sure

File: samples/apps/spring-petclinic/init-database.sh (hunk 7 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/init-database.sh]: the init container is not needed, after all. The only reason for it was that the standard way of initializing db via JPA failed when combined with the autoconfiguration via spring-cloud-bindings and when I wanted the application to run even without the DB... I dropped the idea based on the feedback (that the application is expected to crash with no db) and so now the standard way from spring works and therefore the init container is not necessary

File: samples/apps/spring-petclinic/.minikube/deploy-app.sh (hunk 9 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/.minikube/deploy-app.sh]: this was to be able to run that multiple times for the developement... if you try to expose a deployement as a service that already exists, it would fail... but I changed that - I included the service in the app-deployment.yaml where it is applied, rather then created.

Comment (pmacik) [reply samples/apps/spring-petclinic/.minikube/deploy-app.sh]: so no need for the deletion - removed

File: samples/apps/spring-petclinic/Makefile (hunk 10 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/Makefile:23]: this makefile target is for convenience, no need to have it here, thx

File: samples/apps/spring-petclinic/.patch/pom.xml.patch (hunk 2 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/.patch/pom.xml.patch]: you are right, when I tried the things around I used `spring-data-jpa` and that required additional jdbc dependencies... but it now works without it... no need for these dependencies - removed. Thx.

**Comment by pmacik** [comment]
Sent a pull-request https://github.com/spring-projects/spring-petclinic/pull/887 to enable spring cloud bindings in the petclinic application to avoid the need of patching the app source while building the app container.

File: samples/apps/spring-petclinic/build-app.sh [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (pedjak) [inline samples/apps/spring-petclinic/build-app.sh:1]: should not we wrap this logic inside of a `Dockerfile`, so that we do not need to prepare build environment locally. I think you had this previously?

Comment (pmacik) [reply samples/apps/spring-petclinic/build-app.sh:1]: indeed, I'll do it eventually, but this builds much faster so it allows me going forward faster while I'm still working on it.

File: samples/apps/spring-petclinic/patch/application-postgresql.properties.patch [hunk 12]
```diff
@@ -0,0 +1,4 @@
+diff --git a/src/main/resources/application-postgres.properties b/src/main/resources/application-postgresql.properties
+similarity index 100%
+rename from src/main/resources/application-postgres.properties
+rename to src/main/resources/application-postgresql.properties
```
Comment (pedjak) [inline samples/apps/spring-petclinic/patch/application-postgresql.properties.patch:4]: is renaming really needed?

Comment (pmacik) [reply samples/apps/spring-petclinic/patch/application-postgresql.properties.patch:4]: yes, the spring-cloud-binding requires the type to be `postgresql` and currently it is just `postgres`

Comment (pmacik) [reply samples/apps/spring-petclinic/patch/application-postgresql.properties.patch:4]: well it didn't work before, but it was some number of attempts ago... let me check at the current state

Comment (pmacik) [reply samples/apps/spring-petclinic/patch/application-postgresql.properties.patch]: Ok, no need for that, after all... reverting

Comment (pmacik) [reply samples/apps/spring-petclinic/patch/application-postgresql.properties.patch]: Thx @pedjak!

File: samples/apps/spring-petclinic/build-app.sh (hunk 11 above)
Comment (pmacik) [reply samples/apps/spring-petclinic/build-app.sh]: done

**Review by pedjak — APPROVED** [review]
/lgtm

**Review by baijum — APPROVED** [review]
/lgtm

**Comment by pmacik** [comment]
The `Bind test application to Postgres instance provisioned by Cloud Native Postgres operator` test fails for OpenShift 4.10  because the `Cloud Native Postgres` operator is not available in 4.10 catalog.
